src/H5TSint.c

```diff
@@ -0,0 +1,577 @@
+        HGOTO_DONE(FAIL);
+
+done:
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__init() */
+
+/*-------------------------------------------------------------------------
+ * Function: H5TS_term_package
+ *
+ * Purpose:  Terminate this interface.
+ *
+ * Note:     This function is currently registered via atexit() and is called
+ *              AFTER H5_term_library().
+ *
+ * Return:    void
+ *
+ *-------------------------------------------------------------------------
+ */
+void
+H5TS_term_package(void)
+{
+    FUNC_ENTER_NOAPI_NOINIT_NOERR
+
+    /* Reset global API lock info */
+    H5TS_mutex_destroy(&H5TS_api_info_p.api_mutex);
+    H5TS_atomic_destroy_uint(&H5TS_api_info_p.attempt_lock_count);
+
+    /* Clean up per-thread library info */
+    H5TS__tinfo_term();
+
+    FUNC_LEAVE_NOAPI_VOID
+} /* end H5TS_term_package() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__api_mutex_acquire
+ *
+ * Purpose:     Attempts to acquire the API lock, without blocking
+ *
+ * Note:        On success, the 'acquired' flag indicates if the HDF5 library
+ *              global lock was acquired.
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS__api_mutex_acquire(unsigned lock_count, bool *acquired)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    /* Attempt to acquire the lock */
+    if (H5_UNLIKELY(H5TS_mutex_trylock(&H5TS_api_info_p.api_mutex, acquired) < 0))
+        HGOTO_DONE(FAIL);
+
+    /* If acquired, increment the levels of recursion by 'lock_count' - 1 */
+    if (*acquired) {
+        for (unsigned u = 0; u < (lock_count - 1); u++)
+            if (H5_UNLIKELY(H5TS_mutex_lock(&H5TS_api_info_p.api_mutex) < 0))
+                HGOTO_DONE(FAIL);
+        H5TS_api_info_p.lock_count += lock_count;
+    }
+
+done:
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__api_mutex_acquire() */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS_api_lock
+ *
+ * Purpose:     Increment the global "API" lock counter for accessing the HDF5
+ *              library, acquiring the lock for the thread if the counter is
+ *              initially 0.
+ *
+ * Note:        Multiple (usually recursive) acquisitions of the "API" lock by
+ *              the same thread is permitted with corresponding unlock
+ *              operation(s).
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS_api_lock(void)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_NOAPI_NAMECHECK_ONLY
+
+    /* Initialize the thread-safety code, once */
+    if (H5_UNLIKELY(!H5_INIT_GLOBAL))
+        if (H5_UNLIKELY(H5TS_once(&H5TS_first_init_s, H5TS_ONCE_INIT_FUNC) < 0))
```

Comment:
Is there a way to initialize this package that conforms more to how other packages are initialized? At a minimum we should document this mechanism better. My understanding is that we rely on the FUNC_ENTER_NOAPI at the top of H5_init_library() calling this function before anything else enters this package, is this correct?